README.md
# Slither-Feeder-Bot by l3mpik .
# Slither Client by https://github.com/dahquan
# [[Documentation](http://dahquan.github.io/rattlesnake/)]
# [[Examples](http://dahquan.github.io/rattlesnake/examples.html)]

## 1. Download Slither-Feeder-Bot
## 2. Go to feeder path
## 3. Double click "firstrun.bat" 
## 4. Next run Start Bots.exe
## 5. : )

## Special Thanks To:
- [dahquan/rattlesnake] (https://github.com/dahquan/)
- [iiegor/slither](https://github.com/iiegor/slither)
- [ClitherProject/Slither.io-Protocol](https://github.com/ClitherProject/Slither.io-Protocol)
- [Google](http://www.google.com)
- 
## Social
[![Join the chat at https://gitter.im/dahquan/rattlesnake](https://badges.gitter.im/dahquan/rattlesnake.svg)](https://gitter.im/dahquan/rattlesnake?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)


changelog.txt
#5.27.2016
--- Feeder 
    + Bots Name
	+ Random Skin
    + Little code edit
	
--- Server
    - Delete socket.on('off') <-- idk why i add this xD
	- Little code edit

--  User script
    + Add Skin change 
	+ Add Bots name 

#5.27.2016
--- Feeder
   + Random names
   + Random skin

--- User script
   + Random names
   + Random skin
   
#5.27.2016
--- User script
   + Gui for bots name/random skin (sorry for bad gui im bad in css)
   
--- Server/User script/Feeder
   + Purified code + beautification code
   
FROM: # #5.27.2016

--- XX Times edit feeder

-- XX Times edit userscript

TO:  #6.13.2016
--- Feeder Fix ( 0 bots in console )

--- Bot start when u click Play button

Next: 

--- User skink changer ( rotator / hack ) 

--- Show FPS In left bots menu 

feeder.js
'use strict'

////////////////////////////////////////////////////////////
//                                                       //
//     Feeder/server/userscript by l3mpik			     //
///////////////////////////////////////////////////////////

const Bot = require('./dist');
const express = require('express');
const app = express();
const fs = require('fs');
const path = require('path');
const socket = require('socket.io-client')('ws://127.0.0.1:4000');

process.on('uncaughtException', function(err) {})

let proxies = fs
    .readFileSync(path.join(__dirname, 'proxies.txt'))
    .toString()
    .split(/\r?\n/)
    .filter(function(line) {
        return line.length > 0
    });

let server = '';
let b_name = '';
let skin   = '';
let alive = 0;
let gX = 0;
let gY = 0;

let perProxy = 2;

const bots = [];


function spawn() {

    bots.forEach(function(bot) {
        bot.close()
        console.log('Bot disconnect!');
    });

    alive = 0;
	
	socket.emit('bcount', alive);

    setTimeout(function() {

      
        proxies.forEach(function(proxy, pidx) {
            for (let i = 0; i < perProxy; i++) {
                const bot = new Bot({
                    name: b_name,
                    reconnect: true,
                    skin: skin,
                    server: server
                })

                bot.on('position', function(position, snake) {
                    snake.facePosition(gX, gY);
                })

                bot.on('spawn', function() {
                    alive++;
                    socket.emit('bcount', alive);
					console.log('Available proxy: ' + proxies.length + '\n Chance to spawn max: ' + proxies.length * perProxy + ' bots' + ' Now: ' + alive + '\n\n\n\n\n\n\n\n\n\n\n');
                })

                bot.on('dead', function() {
                    alive--;
                    socket.emit('bcount', alive);
					console.log('Available proxy: ' + proxies.length + '\n Chance to spawn max: ' + proxies.length * perProxy + ' bots' + ' Now: ' + alive + '\n\n\n\n\n\n\n\n\n\n\n');
                })
				
                 bot.on('disconnect', function() {
                    alive--;
                    socket.emit('bcount', alive);
					console.log('Available proxy: ' + proxies.length + '\n Chance to spawn max: ' + proxies.length * perProxy + ' bots' + ' Now: ' + alive + '\n\n\n\n\n\n\n\n\n\n\n');
                })

                bots.push(bot)
                bot.connect(proxy);
				
				  

            }
        })
    }, 1000);

}




socket.on('pos', function(xx, yy) {

    gX = xx;
    gY = yy;

});

socket.on('cmd', function(c) {

    bots.forEach(function(bot) {
        const snake = bot.me()
        if (bot.connected && snake) {

            snake.toggleSpeeding(c === 'on')

        }
    })

});

function r_skin(){
	
setInterval(function(){
	
	i++;
	skin = i;
	if(i == 39) {
		
		i = 0;
		
	}
	
},200);	
	
}

socket.on('server', function(data) {

    server = data[0];
    b_name = data[1];

    if (data[2] == -1) {

        skin = r_skin();
    } else {

        skin = data[2];
    }

    spawn();

});


socket.on('cmd', function(ss) {

    if (ss == 1) {
        bots.forEach(function(bot) {
            const snake = bot.me()
            if (bot.connected && snake) {

                snake.toggleSpeeding(c === 'on');

            }
        });

    }

});


console.log('Waiting for client!');

app.listen(3000)

server.js
var express = require("express");
var app = express();
var path = require("path");
var http = require('http').Server(app);
var io = require('socket.io')(http);

var s_port = 4000;

io.on('connection', function(socket) {
    
	console.log('Connection!');
	
    socket.on('pos', function(x, y) {
       
	   io.emit('pos', x, y);

    });

    socket.on('cmd', function(data) {

        if (data == 1) {
            io.emit('cmd', 'on');
        } else {
            io.emit('cmd', 'off');
        }
		
    });

    socket.on('server', function(data) {

            io.emit('server', data);
			console.log(data);
  	
    });


    socket.on('bcount', function(data) {

        //console.log('Bots: ' + data);	
        io.emit('bcount', data);

    })

});


http.listen(s_port, function() {

    console.log("Server Port: " + s_port);

});

src/bot.js
'use strict'

let Socks = require('socks');

import { client as WebSocketClient } from 'websocket'
import { EventEmitter } from 'events'
import util from 'util'
import { Logger, transports } from 'winston'
import tunnel from 'tunnel'
import SocksProxyAgent from 'socks-proxy-agent'

import Packet from './lib/packet'
import Snake from './snake'
import * as parser from './lib/parser'

import { PROTOCOL_VERSION } from './constants/slither'
import * as messages from './constants/messages'

var mode = 'http';

// Bot
class Bot extends EventEmitter {
  // Options
  // name - snake name
  // server - server address to connect to
  // reconnect - whether to reconnect or not after death or disconnect
  constructor(options) {
    super()

    const { name, server, skin } = options

    this.logger = new Logger({
      level: options.logLevel || 'info',
      transports: [new transports.Console()]
    })

    this.options = options
    this.name = name
    this.server = server
    this.skin = skin

    this.conn = null
    this.connected = false

    this.leaderboard = []
    this.snakes = {}
    this.myId = null
  }

  connecthttp(proxyServer) {
    this.logger.debug(`Connecting bot ${this.name} (${this.server})`.yellow)
    this.proxyServer = proxyServer

    const client = new WebSocketClient()
    const requestOptions = {}

    // Tunnel through proxy server if the option is there
    if(typeof proxyServer === 'string') {
      const AUTH = process.env.PROXY_AUTH || null
	  mode = 'http';
      if(proxyServer.indexOf('socks') === 0) {
        mode = 'socks'
      }

      if(mode === 'http') {
        const proxy = {
          host: proxyServer,
          port: 80
        }

        let idx = proxy.host.indexOf(':')
        if(idx > 0) {
          proxy.port = proxy.host.substring(idx + 1)
          proxy.host = proxy.host.substring(0, idx)
        }

        if(AUTH) {
          proxy.proxyAuth = AUTH
        }

        requestOptions.agent = tunnel.httpOverHttp({
          proxy
        });
	} else if (mode === 'socks5') {
    console.log('Mode socks5, proxy', proxyServer);
    let temp = proxyServer.split(':');
    let socksIp = temp[0]
    let socksPort = temp[1];
    requestOptions.agent = new Socks.Agent({
        proxy: {
             ipaddress: socksIp,
            port: socksPort,
            type: 5
        }
    });
} else if (mode === 'socks4') {
                    console.log('Mode socks4, proxy', proxyServer);
                    let temp = proxyServer.split(':');
                    let socksIp = temp[0]
                    let socksPort = temp[1];
                    requestOptions.agent = new Socks.Agent({
                        proxy: {
                            ipaddress: socksIp,
                            port: socksPort,
                            type: 4
                        }
                    });
                }
      }
    } else if(typeof proxyServer === 'object') {
      requestOptions.agent = proxyServer
    }

    // connectFailed
    client.on('connectFailed', err => {
      this.logger.debug('[%s] %s', this.name, err)

      this.emit('error', err)
      this.emit('errorConnect', err)
    })

    client.on('connect', this.onConnect.bind(this))
    client.connect(`ws://${this.server}/slither`, null, 'http://slither.io', null, requestOptions)
  }
  connectsocks4(proxyServer) {
    this.logger.debug(`Connecting bot ${this.name} (${this.server})`.yellow)
    this.proxyServer = proxyServer

    const client = new WebSocketClient()
    const requestOptions = {}

    // Tunnel through proxy server if the option is there
    if(typeof proxyServer === 'string') {
      const AUTH = process.env.PROXY_AUTH || null
	  mode = 'socks4';
      if(proxyServer.indexOf('socks') === 0) {
        mode = 'socks'
      }

      if(mode === 'http') {
        const proxy = {
          host: proxyServer,
          port: 80
        }

        let idx = proxy.host.indexOf(':')
        if(idx > 0) {
          proxy.port = proxy.host.substring(idx + 1)
          proxy.host = proxy.host.substring(0, idx)
        }

        if(AUTH) {
          proxy.proxyAuth = AUTH
        }

        requestOptions.agent = tunnel.httpOverHttp({
          proxy
        });
	} else if (mode === 'socks5') {
    console.log('Mode socks5, proxy', proxyServer);
    let temp = proxyServer.split(':');
    let socksIp = temp[0]
    let socksPort = temp[1];
    requestOptions.agent = new Socks.Agent({
        proxy: {
             ipaddress: socksIp,
            port: socksPort,
            type: 5
        }
    });
} else if (mode === 'socks4') {
                    console.log('Mode socks4, proxy', proxyServer);
                    let temp = proxyServer.split(':');
                    let socksIp = temp[0]
                    let socksPort = temp[1];
                    requestOptions.agent = new Socks.Agent({
                        proxy: {
                            ipaddress: socksIp,
                            port: socksPort,
                            type: 4
                        }
                    });
                }
      }
    } else if(typeof proxyServer === 'object') {
      requestOptions.agent = proxyServer
    }
	connectsocks5(proxyServer) {
    this.logger.debug(`Connecting bot ${this.name} (${this.server})`.yellow)
    this.proxyServer = proxyServer

    const client = new WebSocketClient()
    const requestOptions = {}

    // Tunnel through proxy server if the option is there
    if(typeof proxyServer === 'string') {
      const AUTH = process.env.PROXY_AUTH || null
	  mode = 'socks5';
      if(proxyServer.indexOf('socks') === 0) {
        mode = 'socks'
      }

      if(mode === 'http') {
        const proxy = {
          host: proxyServer,
          port: 80
        }

        let idx = proxy.host.indexOf(':')
        if(idx > 0) {
          proxy.port = proxy.host.substring(idx + 1)
          proxy.host = proxy.host.substring(0, idx)
        }

        if(AUTH) {
          proxy.proxyAuth = AUTH
        }

        requestOptions.agent = tunnel.httpOverHttp({
          proxy
        });
	} else if (mode === 'socks5') {
    console.log('Mode socks5, proxy', proxyServer);
    let temp = proxyServer.split(':');
    let socksIp = temp[0]
    let socksPort = temp[1];
    requestOptions.agent = new Socks.Agent({
        proxy: {
             ipaddress: socksIp,
            port: socksPort,
            type: 5
        }
    });
} else if (mode === 'socks4') {
                    console.log('Mode socks4, proxy', proxyServer);
                    let temp = proxyServer.split(':');
                    let socksIp = temp[0]
                    let socksPort = temp[1];
                    requestOptions.agent = new Socks.Agent({
                        proxy: {
                            ipaddress: socksIp,
                            port: socksPort,
                            type: 4
                        }
                    });
                }
      }
    } else if(typeof proxyServer === 'object') {
      requestOptions.agent = proxyServer
    }

  // me will return the snake that belongs to you
  me() {
    return this.snakes[this.myId]
  }

  // onConnect gets called once we successfully connect to the slither.io
  // game server
  onConnect(connection) {
    this.logger.debug(`[${this.name}] has successfully connected`.rainbow)
    this.conn = connection
    this.connected = true

    connection.on('error', error => {
      this.logger.debug('[%s] Error - %s', this.name, error)
      this.emit('error', error)
    })

    connection.on('close', this.onClose.bind(this))
    connection.on('message', this.onMessage.bind(this))

    // Send the initial packet
    // SetUsernameAndSkin
    // https://github.com/ClitherProject/Slither.io-Protocol/blob/master/Protocol.md#packet-setusernameandskin
    const initialPacket = new Packet(3 + this.name.length)
    initialPacket.put(115)
    initialPacket.put(PROTOCOL_VERSION - 1)
    initialPacket.put(this.skin || 44)
    initialPacket.putString(this.name)
    initialPacket.send(this.conn)
  }

  // onMessage
  onMessage(message) {
    if(message.type !== 'binary') {
      return
    }

    const parsedMessage = parser.parse(message.binaryData)
    if(!parsedMessage) {
      return
    }

    let snake;

    switch(parsedMessage.messageType) {
      case messages.INITIAL:
        // Make sure the protocol version matches
        if(parsedMessage.protocolVersion !== PROTOCOL_VERSION) {
          this.logger.error('[%s] Protocol version mismatch. %d > %d', this.name,
            PROTOCOL_VERSION, parsedMessage.protocolVersion)
          this.conn.close()
        }
        break

      case messages.SNAKE:
        if(parsedMessage.removing) {
          if(typeof this.snakes[parsedMessage.id] !== 'undefined') {
            const snake = this.snakes[parsedMessage.id]

            this.logger.debug('[%s] Removing snake %s', this.name,
              snake.name)
            this.emit('snakeLeft', snake, (snake.id === this.myId))

            delete this.snakes[parsedMessage.id]
          }
          return
        }

        snake = new Snake(this, parsedMessage)

        // The first snake in our view will be us
        if(!this.myId) {
          this.myId = parsedMessage.id
          this.emit('spawn', snake)
        }

        this.snakes[parsedMessage.id] = snake
        this.emit('snakeEnter', snake, (parsedMessage.id === this.myId))
        break

      case messages.LEADERBOARD:
        this.leaderboard = parsedMessage.snakes
        this.emit('leaderboard', this.leaderboard)
        break

      case messages.DEAD:
        this.logger.debug('[%s] Just died', this.name)
        this.emit('dead', parsedMessage)
        break

      case messages.EAT:
        this.emit('ate')
        break;

      case messages.POSITION:
        snake = this.snakes[parsedMessage.id]
        if(typeof snake !== 'undefined') {
          snake.x = parsedMessage.x
          snake.y = parsedMessage.y

          this.emit('snakePosition', parsedMessage, snake)

          if(snake.id === this.myId) {
            this.emit('position', parsedMessage, snake)
          }
        }
        break
    }
  }

  // onClose
  onClose(reason) {
    this.logger.debug(`[${this.name}] Disconnected (${reason})`.red)
    this.connected = false
    this.conn = null
    this.leaderboard = []
    this.snakes = {}
    this.myId = null

    this.emit('disconnected')

    // Reconnect after we close (probably died)
    if(this.options.reconnect) {
      setTimeout(() => {
        this.connect(this.proxyServer)
      }, 2000) // TODO: Make this a configuration
    }
  }

  // close
  close() {
    if(this.conn !== null) {
      this.conn.close()
    }
  }

  // toJSON
  toJSON() {
    const { id, name, server, connected, leaderboard } = this
    return {
      id,
      name,
      server,
      connected,
      leaderboard,

      // Will deal with this later
      enabled: true
    }
  }
}

// For some reason a ton of slither.io servers don't return the expected WebSocket
// accept key during handshake. This causes the bot to terminate the connection.
//
// This "Hack" will stop it from closing
//
// TODO: Figure out the cause of this
WebSocketClient.prototype.failHandshake = function(reason) {
  if(reason.indexOf('Sec-WebSocket-Accept header from server didn\'t match') >= 0) {
    // Ignore it and accept
    this.succeedHandshake()
  }
}
export default Bot


userscript.js
// ==UserScript==
// @name         Slither Feeder Bot
// @namespace    Slither Feeder Bot  by l3mpik
// @version      1.0
// @description  Slither Feeder Bot
// @author       l3mpik
// @match        *://slither.io/*
// @require      https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js
// @require      https://cdnjs.cloudflare.com/ajax/libs/socket.io/1.4.5/socket.io.min.js
// @grant        none
// ==/UserScript==
//====================   TRY :)   ==========================//
var skin = -1; // -1 <-- Random Skin 1 <-- User ur skin

//======================================================//

var vps = 0;

var updatespeed = 100;

var st_click = 0;

var _ip = "";

if (skin != -1) {
    data[2] = localStorage.snakercv;
}

var _data = [];

_data[1] = "slither-feeder-bot";
_data[2] = skin;

//======================================================//


$('iframe').remove();

$("body").append("<div id='b_menu' style='width: 350; height: 120; background-color: #000000; opacity: 0.7; top: -15%; border: 0.5px solid #ffffff; position: absolute; color: #ffffff; a { font-size: 9px; }'><a>Bot Name: </a><input id='bname' type='text' placeholder='Bot name'> <br>Bot skin: <input type='text' id='bskin' placeholder='Skin 1-39 or -1'> <button id='save' style='width: 150px; height: 25px; background:#4dff4d; border: 0px; border-radius: 5px;'>Save</button>");

$("canvas:eq(3)").after("<div style='height: 250px; background-color: #000000; opacity: 0.8; zoom: 1; width: 205px; top: 1%; left: 1%; display: block; position: fixed; text-align: center; font-size: 15px; color: #ffffff; padding: 5px; font-family: Ubuntu; border: 0.5px solid #ffffff; border-radius: 5px;'> <div style='color:#ffffff; display: inline; -moz-opacity:1; -khtml-opacity: 1; opacity:1; filter:alpha(opacity=100); padding: 10px;'><a>Bots</a></div> <div style='color:#ffffff; display: inline; opacity:0.8; filter:alpha(opacity=100); padding: 10px;' position: fixed;><br><a>Status:</a><a id='count' > Off </a> </div> <div style='color:#ffffff; display: inline; -moz-opacity:1; -khtml-opacity: 1; opacity:1; filter:alpha(opacity=100); padding: 10px;'><br><a></a> Move To Head: <a id='moveh' >On</a> </div> <div style='color:#ffffff; display: inline; -moz-opacity:1; -khtml-opacity: 1; opacity:1; filter:alpha(opacity=100); padding: 10px;'><br><a>X</a> - Snake Speed: <a id='isspeed' >Off</a> <br><button id='start' style='width: 150px; height: 25px; background:#ff3333; border: 0px; border-radius: 5px;'>OFF</button><br>MODE: <font color='#00ff00'><a id='mode' ></a></font></div> ");

console.log('Add GUI!');

$('#b_menu').mouseenter(function() {


    $('#b_menu').animate({

        top: "0%",
        border: "2px solid #ff0000"
    });

});

$('#b_menu').mouseleave(function() {


    $('#b_menu').animate({

        top: "-15%"

    });

});

$('#fsrv').click(function() {

    if (!snake) {

        console.log('Snake die');

    } else {
        if (bso.ip != undefined || null && !snake) {
            var ip = bso.ip;
            var port = bso.po;
            connect('ws://' + ip + port);
        }
    }

});

if (vps == 1) {
    var socket = io.connect('ws://188.68.252.227:3333');
    $("#mode").text('VPS');
} else {
    var socket = io.connect('ws://127.0.0.1:3000');
    $("#mode").text('LOCAL');
}


window.onkeydown = function(event) {
    if (event.keyCode === 71) {
        $("#botlayer").fadeToggle(1000);
    }
};

document.body.onmousewheel = zoom;

window.onkeydown = function(e) {

    if (e.keyCode === 88) {
        socket.emit('cmd', 1);
        $('#isspeed').text('On');
    }
}

window.onkeyup = function(e) {

    if (e.keyCode === 88) {
        socket.emit('cmd', 0);
        $('#isspeed').text('Off');
    }
}

function zoom(e) {

    gsc *= Math.pow(0.9, e.wheelDelta / -120 || e.detail || 0);

}

function rsk() {


    var rsin = setInterval(function() {


        function gr() {

            var rn = Math.floor((Math.random() * 39) + 1);

            return rn;
        }

        setSkin(snake, gr());


    }, 150);

}

var spos = setInterval(function() {

    if (window["snake"] !== undefined) {

        if (snake != null) {
            var x = snake.xx;
            var y = snake.yy;

            socket.emit('pos', x, y);
        }

    }

}, updatespeed);


$('#bname').change(function() {

    var name = $('#bname').val();

    _data[1] = name;

});

$('#bskin').change(function() {

    var skin = $('#bskin').val();

    _data[2] = skin;

});

$("#playh").click(function() {

    lbh.textContent = "Slither-Feeder-Bot";
    ii.src = "";

    if (window["bso"] !== undefined) {

        _data[0] = "" + bso.ip + ":" + bso.po;

    }

});

$("#start").click(function() {
    if (st_click == 0) {
        st_click = 1;

        $("#start").css('background', '#4dff4d');
        $("#start").text('Waiting...');

        setTimeout(function() {

            $("#start").text('On');

            if (window["bso"] !== undefined) {

                _data[0] = "" + bso.ip + ":" + bso.po;

                socket.emit('server', _data)

            } else {
                console.log('Try refresh page : /');
            }

        }, 2000);

    } else {
        st_click = 0;

        $("#start").css('background', '#ff3333');
        $("#start").text('OFf');
    }

});

socket.on('bcount', function(data) {

    $('#count').text(data);

});